Allow custom field formatting

hanak opened the pull request:
**Problem**
EPICS DB format allows for constants in place of links. In case of the record type lsi, this can be used to set value of the record. Since a link is a string, any string is processed as a link unless it is marked as a constant array (e.g. `["constant"]`) (see function dbParseLink(), epics_base/modules/database/src/ioc/dbStatic/dbStaticLib.c). However, epicsdbbuild assumes that a field value has to be decorated by quotes.

**Solution**
Since epicsdbbuild allows custom validation if a value has an attribute "Validate", we can add a similar mechanism. If a value has an attribute "Print", epicsdbbuild will call the function Print() in order to format value. This enables occasional customizations without a need to extend the framework.

**Test**
Add test example into the list of examples.

hanak commented:
Hello,

I have encountered this problem while trying to set lsi with a constant value. Using `field(VAL, "value")` was rejected by EPICS DB. Yet, `field(INP, ["value"])` worked. I am not able to tell whether I am using a gap in EPICS or not. Thus, I converged towards extending hooks/customization rather than introducing an option to epicsdbbuild for constant links. Maybe, epicsdbbuild will get one in future. Even if that happens, the proposed solution should not be affected.

Would it be possible to review this pull request, please?

Thank you.

Sincerely,
Ivo Hanak

Araneidae commented:
This is an interesting hook.  Have to say I've not seen this syntax for values, I guess this must be new to EPICS v7?  Is this syntax documented anywhere?

It's quite alarming to me that you're reporting `field(VAL, "value")` is rejected by EPICS; this sounds like something that needs a careful fix.  I guess we'd need to know somehow that this particular instance of `VAL` needs `[...]` quotes rather than the usual `"..."` quotes.

Have to confess I'm reluctant to take this patch until we understand the problem we're solving better.  It does sound like the problem might not be with the value.

I'm intrigued by your note about proper escaping in the patch.  You're quite right that this is a missing feature.  I seem to remember when I first thought about this throwing my hands up because there is no reliable quotation mechanism!  Maybe my memory is unfair, and this would be worth fixing.

hanak commented:
Thank you for your response. I am not sure whether and how EPICS documented this syntax in wherever their official copy of documentation is expected to be. Nevertheless, EPICS 7.0.4.1 mentioned this functionality in section "Constant Link Values". Yet, I am using EPICS 7.0.3.1 and it works there because I have quotation mark inside the brackets. According to the code (function dbParseLink()), a comma would to the same trick. This form has been there for 5 years and it was modified to get EPICS 7.0.4.1 behavior.

Regarding the VAL field. Let me expand the description a little bit. My goal was to initiate a record of type **lsi** with a constant string. Therefore, I tried following:
```
record(lsi, "myRec") {
  field(SIZV, "256")
  field(VAL, "initial value")
}
```
However, EPICS DB loader generated following message:
`Can't set "myRec.VAL" to "initial value" Error (512,512)`

The message contains an defect ("Error (512,512)" should be "Error (512,11)") and I have reported it already together with a potential solution (https://bugs.launchpad.net/epics-base/+bug/1907077).

As far as it looks like, the error is caused by the fact that the field VAL of lsi is of type DBF_NOACCESS. This type is not recognized by the processing function dbPutString(). Fortunately, I can use the fact the INLINK can be a constant.Thus, if I supply a constant, I will get desired result:
```
record(lsi, "myRec") {
  field(INP, ["initial value"])
  field(SIZV, "256")
}
```

Since the bracket syntax is not use massively, I proposed a hook which could enable use of new and/or experimental extensions to EPICS without a need to modify epicsdbbuilder. Alternatively, an array could be the indicate the brackets are meant to be used (`myRec.INP = ['my const']`).

Regarding the quotes. I think `\"` should do the trick. At least asLib_lex.l seems to suggest that anything following a backslash is considered an escaped content. Definitely, one need to confirm it. Still, the code does not escape the string right now. If anyone tried to work around it by escaping the string prior making a call to epicsdbstring, adding escaping would cause them to experience moments of intensive sadness. Nevertheless, this is slightly a different problem than the problem I am trying to address here. The comment there is meant to say: "yes, the code is aware of the problem".

hanak commented:
I just realized, I have not asked about it before but which version of EPICS do you use? I am using EPICS 7.0.3.1 but I plan to upgrade to EPICS 7.0.4.1.

Araneidae commented:
We're very behind the times at Diamond at the moment: we're on EPICS 3.14.12, though I do have colleagues who are doing work on EPICS 7, but I know little about it.

Thank you for the documentation link.  There seem to be two main points:

> In this release, constant links can be expressed using JSON array syntax and may provide array initialization of values containing integers, doubles or strings

and, further up the document:

> The new link types are documented in a separate document that gets generated at build time and installed as `html/links.html`

and I see reference to a "new CONSTANT field type".

Whatever we do needs to be backwards compatible with EPICS 3.14, so I guess my first question has to be: can we tell statically (using dbStatic functions) whether a field is of the new CONSTANT type?

Hmmm.  It looks as if string quoting *does* work.  I think I'd rather fix that and warn in the readme and documentation that this is an incompatible change.  Anyhow, this would be separate from your pull request, I've created issue #9 for this.

I'll write a separate response to the specifics of this pull request.

Araneidae commented:
I think I have two reservations about this pull request:

1. I'm not happy that we're creating a mechanism for injecting arbitrary text into the middle of a generated file.  I know that this is required as a workaround, but I'd prefer to get a bit closer to fixing the underlying problem.
2. I'm not convinced that tying the fix to the *value* is the correct approach.

Am failing to come up with a sensible alternative, though.  Perhaps a start would be to rename `Print` to something less generic like `FormatEpicsValue`.  Think I'd like to see some thoughts on whether we can fix the problem more generally.

I guess your starting point is the fact that when you write `rec.INP = 'initial value'` you really want to be able to write something like `rec.INP = ['initial value']` and have it work.  So one approach might be to have some special value specific formatting code that treats lists specially.  This would inevitably be ugly with chained`isinstance` checks.

Let's try and drive this by application.  I think your test already shows that the hook is incomplete: what you're having to write when exercising this feature is already too clunky for normal use: maybe the custom field support should be integrated into the change with a special field wrapper type exported by `recordbase.py`?

hanak commented:
Thank you for pointing out on the documentation generated through the build. I have habit of looking at the web for documentation. My mistake. I am sorry.

> Whatever we do needs to be backwards compatible with EPICS 3.14, so I guess my first question has to be: can we tell statically (using dbStatic functions) whether a field is of the new CONSTANT type?

I think the the proposed solution tries to work around the question. Yet, if we were looking at it in an attempt to get some consistent behavior, I am not sure whether there is a straightforward answer. Let me illustrate on a result of a quick search through source code of EPICS 7.0.4.1+. Let our journey start at a function dbPutString() (dbStaticLib.c) because dbPutString() is used while processing the field (dbRecordField() defined in dbLexRoutines.c and referenced by dbYacc.y):
1. dbPutString() tries to check field type. If the field type is DBF_INLINK, DBF_OUTLINK, or DBF_FWDLINK, it will make a call to a function dbParseLink() (line 2602).
2. The function dbParseLink() will process the supplied string. At one specific moment (line 2334) it will check whether the string starts with "[" and ends with "]". If so, it will report its type as CONSTANT and return back to dbPutString().
3. dbPutString() will check the type identifies by dbParseLink(). If it is CONSTANT, it will process the string as constant.

This means, a field of type DBF_INLINK should be able to accept an array of strings. Yet, it is still DBF_INLINK. Maybe, the way out is the use the function dbParseLink() to confirm validity. If it parses the string representation, it is a valid link. Unfortunately, this will be able to deal with parameters as parameters are processed later. A solution could be if the user supplied an example value of the parameter while declaring it (e.g. `p = Parameter("P", "Port", example="mydevice")`). Optionally, it would use the default if supplied. Then, we could build the link and pass it through dbParseLink().

> Let's try and drive this by application. I think your test already shows that the hook is incomplete: what you're having to write when exercising this feature is already too clunky for normal use: maybe the custom field support should be integrated into the change with a special field wrapper type exported by recordbase.py?

I think the example should have avoided the function `Validate`. I think epicsdbbuilder confirms it is there before it makes the call. I agree, the example is more complex that it should be. Yet, let me continue with your paragraph. Does it mean following?
1. Define a class `CustomFieldValue` inside `recordbase.py`.
2. Implement two methods of the class: `Validate` and `FormatEpicsValue`. By default, it would return `True` on `Validate` and properly quoted string version of the value on `FormatEpicsValue`.

Did I get it right?

Araneidae commented:
I think I have what might be simpler idea to address the problem you're trying solve today.  If the value is an instance of a list (more tricky than meets the eye to check, more on this below) the format it as a list of strings with `[` ... `]` brackets containing quoted values, otherwise format as a single quoted string.

Maybe something as simple as:
```python
def format_value(value):
    if isinstance(value, list):
        return '[' + ','.join(map(format_string, value)) + ']'
    else:
        return format_string(value)

def format_string(value):
    return '"' + ''.join(map(quote_char, value)) + '"'

def quote_char(ch):
    if ch in '"\\':
        return '\\' + ch
    else:
        return ch
```

hanak commented:
Ok. Actually, while I was reading `dbYacc.y` I noticed they use `json_value` in place of the field. Maybe, the be solution would be to threat the input as json and let Python to format it as such.

Araneidae commented:
I've just pushed a fix to the quoting issue.

I was wondering where the special `Validate` hook comes from, turns out it's required to implement the `Parameter` value which completely bypasses validation.

Are we going to be able to validate list values?   I'd be happy to support special treatment of `list` type values, but can we reliably validate them?  I feel there is a can of worms here that I don't properly understand.

hanak commented:
I agree. There is complexity in there. Maybe, this is up to the user of the package. The package can just allow the user to extend validation which fits the use case and offer a straightforward way how one can update generated content instead of trying to fit all. In my case, I have added a concept of a placeholder element. This concept is specific to my use of epicsdbbuilder. Still, epicsdbbuilder allows me to do so without major obstacles. I find this quite useful.

Yet, let us get back to the case of constant links. It looks like we have following options:
1. Add a hook `FormatEpicsValue` similar to `Validate` and let the user to take care of formatting and validation (if needed).
2. Add special handling of arrays (`quote_string(str(value)) if not isinstance(value, list) else format_list(value)`) and let epicsdbbuilder to deal with formatting and some level of validation.

As you have mentioned, the validation is complex. It might be too complex. Unnecessarily complex. Maybe, it is not worth to think about it and move the responsibility to the user (i.e. the option 1). If there are going to be changes to accommodate EPICS 7.0.4+, the option 1 can still co-exist without unwanted interaction.

What do you think?

hanak commented:
I am sorry for being impatient but which option would you prefer? It feels to me you see validation of array members as a challenge. Did I get it right?

Araneidae commented:
Hi @hanak .  Sorry for the radio silence, am now back from Christmas and New Year break.

I think there are two choices here:
1. "Magically" choose the correct representation using metadata from the static DB and the specified datatype.
2. Use a special formatting wrapper type.

I would prefer the "magic" solution (it's the Python way, after all) if we could make it work ... but I don't know how to do it, and we risk generating unexpected unusable .db files, so I think we're back to a wrapper type.

My problem with your solution is, I think, simply that it's incomplete.  Rather than just adding a magic hook (your `Print`, though actually I'd prefer we called it more like `Format` or even `FormatDb`) and asking users to create their own wrappers and implementations of `Print`/`Format`, I suggest that you add a wrapper class which does the work.

So, I think the solution is a new class, let's call it `ConstArray`, so that you can write your record as
```python
ai.record(record_name, INP = ConstArray('link'))
```
The big difference is this: the special attribute is an undocumented internal implementation detail (just as the `Validate` and `ValidateLater` attributes are), whereas this new `ConstArray` class will be a documented way to write what you want.

Does this make sense?

hanak commented:
Is `ConstArray` going to implement `Validate`? If so, should it just confirm that none of the elements are DB records or should it be more complex?

Araneidae commented:
It'd be good to do some validation: I suggest you validate as much as you sensibly can!  Unfortunately I'm not familiar with the new .db format, so you'll need to find out just what you _can_ validate.  Ideally the generated .db file is guaranteed to be valid, but we're limited by the available API.  Do what you can...

hanak commented:
I have updated the pull request:
* Constant link array is supported through a class ConstArray. The class is defined in a new file const_array.py
* ConstArray will check assigned content upon assigning. It will trigger an assert if:
  * the content is not a list or a tuple
  * any element of the tuple is not a number, a string, or a parameter
  * some elements are strings/parameters and others numbers
* Behavior of ConstArray is covered through automated unit tests. Also, I have added a basic test for quite_string. That test should be extended with more special cases. I think this can be done gradually every time we run into some issue. I have not checked it but maybe GitHub offers some "free" minutes of runtime for a public project. If so, we might want to try to run those tests for every pull request automatically. Yet, I would consider enabling this behavior through a different pull request.

Would it be possible to review the pull request, please?

Araneidae commented:
Will do.  Thanks for your efforts!

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+        numbers = False
+        strings = False
+        for index, value in enumerate(value):
+            assert value is None \
```
`None` is not a valid value, and this occurs in lots of other places.

hanak replied:
The intention with `None` as a value in array was to skip a value without a need for a filtering the array. Removed.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+        self._assert_valid(value)
+        self._value = value
+
+    def _assert_valid(self, value, record=None, fieldname=None):
```
What are the `record` and `fieldname` arguments for?  They're never assigned.

hanak replied:
The parameters were meant to be set in `Validate`. Corrected.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+            return False
+
+    def FormatDb(self, record, fieldname):
+        if self._value is None or len(self._value) == 0:
```
`None` should not be a valid value

hanak replied:
Removed.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+            self._format_constant(v)
+            for v in self._value if v is not None]
+        if len(formatted) == 0:
+            return ''
```
I'm pretty certain that an empty field is not valid in a `.db` file.  This is going to generate something like
```
field(NAME, )
```
which I'm sure won't parse correctly.

hanak replied:
Good point. I have replaced it with `[""]` (empty string).

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+
+
+class ConstArray:
+    """Constant Link Values. EPICS Base 7.0.4.1 and above.
```
This documentation would be useful in `api.rst`, which at the moment is disappointingly bare.

hanak replied:
The intention was to favor IDE (with intellisense) over static text. Among others, a comment next to the code has higher chance for receiving updates rather than a static document. Yet, I recognize it depends on overall style of the documentation. Do you want me to duplicate the information in `api.rst` or move the information there leaving something much shorter here?

Araneidae reviewed (changes requested):
I have quite a few scattered points, but I have one big problem: I would like to be more confident that `.db` files generated using `ConstArray` will actually parse successfully when loaded.  Unfortunately I don't think you're doing anything to check this.

Does `dbVerify` work in the version of EPICS you're using?  If it's still broken (or missing) then I guess my point is moot and I'll have to accept that there's no point in my grumbling about weak validation.

Finally, `None` is not a valid field value: it is used to delete fields, and so shouldn't be supported as part of `ConstArray`

hanak commented on docs/api.rst:
```diff
@@ -217,6 +217,11 @@ Helper Functions
+..  class::
+    ConstArray(array)
+
+    Used for constant link values available in EPICS 7.0.4.1.
```
I think you are correct. I made assumptions that the Release Notes of EPICS 7.0.4.1 (https://epics.anl.gov/base/R7-0/4-docs/RELEASE_NOTES.html) is about EPICS 7.0.4.1 only without reading every line of the document. This assumption was incorrect as the document contains information about previous releases, too. I this case, the section I am looking at shows a title "Changes made between 3.16.0.1 and 3.16.1". Now, the text of the paragraph seems to indicate the change in "3.16.1" is about enabling JSON-like array syntax. It states the functionality is older and it used be known as "constant links". Sadly, the document does seem to mention it again so official syntax is not clear just from this document.

Given the fact we generate JSON-like syntax, it feels "3.16.1" should be the the answer here. Would this be sufficient justification?

hanak commented on epicsdbbuilder/__init__.py:
```diff
@@ -17,4 +17,4 @@ def ExportModules(globs, base_module, modulenames):
 __all__ = ExportModules(globals(),
     'epicsdbbuilder',
-    ['dbd', 'recordbase', 'fanout', 'recordset', 'recordnames', 'parameter'])
+    ['dbd', 'recordbase', 'fanout', 'recordset', 'recordnames', 'parameter', 'const_array'])
```
Given the fact this is __init__.py rather then a regular file, would you mind if I add line breaks leaving a single name per line and order it alphabetically? I think it can make it more readable and I need to change the line regardless the style.

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+    def Validate(self, record, fieldname):
+        try:
+            self._assert_valid(self._value)
+            return True
```
You are correct. I think I got confused by Validate in the class Parameter (parameter.py). There, it uses following:
```
    def Validate(self, record, field):
        return True
```
Is this the the same "Validate" as ConstArray's "Validate"? If so, can I replace "return True" with "pass" here, too?

Araneidae replied:
Good point, I think you're right.

hanak replied:
Code (paramers.py) has been updated accordingly.

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+        self._value = value
+
+    def _assert_valid(self, value, record=None, fieldname=None):
+        id = 'ConstArray' \
```
I am not sure I got the comment right. Yet, I am going to assume you are referring to the method `Validate`. In such a case, the method `_assert_valid' is used by the setter of a property `value`. There, I have no record and no fieldname. Definitely, I can move "None" there to avoid the default values here or avoid the check until the DB is generated. I just thought checking it early might be more convenient. What would you prefer?

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+                or isinstance(value, int) \
+                or isinstance(value, float) \
+                or isinstance(value, bool) \
+                or isinstance(value, Decimal), \
```
I will update the code accordingly.

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+            return '1' if value else '0'
+        return str(value)
+
+    def Validate(self, record, fieldname):
```
Right now, `Validate` will try to check combinations which are not considered valid according to documentation. It does not check versions of EPICS. Yet, this can be extended. Since we are running external code which is not maintained together with EPICS, the only check which may guarantee the behavior is to run the EPICS DB parser in `Validate` assuming EPICS will actually fail rather than skip the input silently. However, `Validate` is run before formatting (which we need for the EPICS DB parser) happens. Thus, it feels inappropriate here. I am afraid, I will need to ask for the intention of this comment. What am I missing?

hanak commented:
Thank you for the review. I have updated the code and added comments to the questions. I think the major point is the use of `dbVerify`. It looks like that it was removed in 3.16.1 and restored in 7.0.1.1. Now, this function returns True in epicsdbbuilder in order to mask the lack of this function in 3.16.1. It makes perfect sense to restore the function (or attempt to) for EPICS 7.0.1.1 (and above). Since epicsdbbuilder is using the function already, it is a good idea to append to this change?

hanak commented:
It looks like code of `dbVerify` is just about parsing a single "value" (a number, an enum, a string). If one attempts to supply to parse INLINK, OUTLINK, and FWDLINK (which is this case), it will just return NULL. And, this is treated as success (according to dbStaticTest.c). As a consequence, `dbVerify` is not useful here.

Now, I tried to see whether I use other parsing routines. It looks like `dbParseLink` is able to parse something but it does not parse content between "[]". It just detects a presence of either `,` or `"` in EPICS 7.0.3.1. This is not correct because it treats `[1]` as a link. Now, this changed in the current EPICS branch "7.0". Anything between "[]" is treated as a constant now. Thanks to difference between EPICS versions, `dbParseLink` could be used to check at least something. Unfortunately, `dbParseLink` requires an instance of a structure `dbLinkInfo` which is available only in the header file. There is no equivalent to `dbAllocBase` or `dbAllocEntry`. This means, we need to use a caller of `dbParseLink`: `dbPutString`.

`dbPutString` assumes that `DBENTRY::pfield` is initialized to an instance of `DBLINK`. Yet, this does not seem to happen with just DBDs being loaded (my guess only). And, loading records requires a call to `x_registerRecordDeviceDriver` or it fails otherwise. Have I interpreted your request correctly?

Araneidae commented:
I'm inclined to accept this merge without proper database validation, but immediately raise an issue so that we can think about ways of properly validating at some point in the future.

I'm happy with the changes you've made, but there are still a lot of uses of `None`.  Think I'll close out the current review (if I can) and start a new one to catch all the `None`s that shouldn't be there (they are numerous, I'm afraid).

Araneidae reviewed (changes requested):
The use of `None` is inconsistent, and I'm going to suggest removing it from _all_ occurrences in `const_array.py`.

In fact, let's start with that.  Simply remove every occurrence of `None` from `const_array.py` (or justify the occurrences you really really want to keep).

hanak commented:
Thank you for your suggestion and patience. I have removed all instances of the keyword `None` from the file `const_array.py`.

hanak commented:
Thank for your comment. Some tests might be removed, yes. . Yet, let me try to add details. I am going to use EPICS 7.0.3.1 for illustration:
1. `[]` will load without any warning or error. Yet, `dbpr` will reveal the INP is `CA_LINK [] NPP NMS`. I would not say this is intended result. This is caused by a function `dbParseLink` (`dbStaticLib.c`) which will identify the content as a link rather than a constant. Unfortunately, this function went through some evolution across versions (at least EPICS 7.0.3.1 differs from the current branch 7.0 at the EPICS launchpad repository).
2. `[""]` will load without any warning or error if the record is `stringin`. `ai` will create a warning `dbConvertJSON: String provided, numeric value(s) expected` (with no hint about which line is the cause). Yet, `dbpr` will show INP as `CONSTANT [""]` for both `stringin` and `ai`.  And, `caget` will return 0 in case `ai`. Still, there is that warning which you will not accept I assume.

So, what if I do following?
1. Block an empty array as not allowed.
2. Call to `dbVerify` even though it does nothing (now) in case of input links. Maybe, it will get better in future.

Araneidae commented:
We can't call `dbVerify` with the content of a `ConstArray`: `dbVerify` takes a single string as argument, we have an array in our hands.  Is there any way to at least statically check that the field we're writing to can take an array as argument?

If empty arrays don't make sense to EPICS (sounds likely enough to me) then, yes, block it as part of the verify step by adding an `assert value` after checking that it is a `list` or `tuple`.

I might be tempted to replace the `isinstance(value (list, tuple))` test with a simple `value = list(value)` assignment.  This would have the virtue of allowing a generator to be passed and is unfussy, though occasionally surprising: calling `ConstArray('str')` would then generate `["s", "t", "r"]` which might be a bit unexpected.

hanak commented:
> We can't call `dbVerify` with the content of a `ConstArray`: `dbVerify` takes a single string as argument, we have an array in our hands. Is there any way to at least statically check that the field we're writing to can take an array as argument?

I agree. Therefore, I considered following sequence:
1) Perform all checks so that formatting can run safely
2) Format `ConstArray` to EPICS DB string
3) Make a call to `dbVerify`

The only check I can think of is about type of the field. I think `ConstArray` makes sense in case of `DBF_INLINK`. The code of `dbStaticLib.c` seems to suggest `DBF_FWDLINK` and `DBF_OUTLINK` might work, too. Yet, I am not sure how this gest interpretted. The field type is stored in a structure `DBENTRY`. However, epicsdbbuilder treats this structure as `void*`. This makes perfect sense as this shields epicsdbbuilder from change inside EPICS. Unfortunately, EPICS does not provide getters which would allow reading the structure. So, I am not able to get the field type easily. Therefore, I was thinking about `dbVerify` which makes checks based on field type. While the function does nothing in case of links, it should block inappropriate assignments.

> If empty arrays don't make sense to EPICS (sounds likely enough to me) then, yes, block it as part of the verify step by adding an `assert value` after checking that it is a `list` or `tuple`.
Yes.
- [ ] Block use of empty `list`/`tuple`. 

> I might be tempted to replace the `isinstance(value (list, tuple))` test with a simple `value = list(value)` assignment. This would have the virtue of allowing a generator to be passed and is unfussy, though occasionally surprising: calling `ConstArray('str')` would then generate `["s", "t", "r"]` which might be a bit unexpected.
I think I would be quite surprised if `'str'` turned to be treated as an array of characters instead. I would not expect it. What other such cases are there? If there are just few, we can just check them and use `list(value)` once we are sure usual conflicting cases were filtered before.

Araneidae commented:
I think `str` is the only sensible alternative case, so how about:
```python
if isinstance(value, str):
    value = [value]
else:
    value = list(value)
```
Simple and to the point!

I'm not sure that `dbVerify` would be able to sensible parse a list rendered as a string, after all, how can it tell whether a string of the form `["str"]` is meant to represent a list or a string?

I suggest we wrap up the changes we can do without calling `dbVerify` so we can merge this, and then focus on looking at what `dbVerify` actually does and how much it can do for us.  Maybe we can try and push some fixes upstream to EPICS base?

hanak commented:
Thank you for your comment. About the code. The proposed code makes `ConstArray("0")` and `ConstArray(["0"])` equivalent. This make sense. Yet, `ConstArray(0)` will fail with an error `'int' object is not iterable`. I can see it might be intentional but I can see it might look like inconsistency between a string ("0" succeeds, ["0"] succeeds) and a number (0 fails, [0] succeeds). Is it intentional?

> I'm not sure that dbVerify would be able to sensible parse a list rendered as a string, after all, how can it tell whether a string of the form ["str"] is meant to represent a list or a string?

I think a field type is the key. `dbVerify()` use is already and it could mimic the rest of the EPICS DB routines. While loading EPICS DB, `dbPutString()` will call `dbParseLink()` if field type is a INLINK, OUTLINK, or FWDLINK. `dbParseLink()` will use the presence of [] in order to tell a string and a string constant apart (see dbStaticLink.c, line 2325). I have not tried to evaluate whether `dbVerify()` may call `dbParseLink()` to get the effect. Yet, this is outside scope of this item, I think.

> I suggest we wrap up the changes we can do without calling dbVerify so we can merge this, and then focus on looking at what dbVerify actually does and how much it can do for us. Maybe we can try and push some fixes upstream to EPICS base?

Yes. Lets do it. I am going to update the code and automated tests once I know whether the behavior around "0", ["0"], 0, and [0] is intentional.

Araneidae commented:
Good point about `ConstArray("0")` vs `ConstArray(0)`.  Neither choice is great, though I suppose it would be more consistent to forbid `str` ...
```python
assert not isinstance(value, str), 'this seems a bit arbitrary'
value = list(value)
```
On the whoke am leaning towards treating `str` as a list of characters: it's simpler and slightly more honest.

hanak commented:
Alright. Let's treat `str` as a list of characters. I update updated the solution accordingly. Also, I have reorganized unit tests with focus on behavior. It should be easier to see result of a given argument.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+            value_list = list(raw_value)
+        except TypeError:
+            value_list = None
+        assert value_list is not None, \
```
I definitely don't approve of this one!  What you've done here is to catch the exception, set a special sentinel value, and then test for the sentinel to finally raise the exception you want to.  A more honest way to write this would be:
```python
try:
    value = list(value)
except TypeError:
    raise AssertionError('blah blah')
```

hanak replied:
I am going to update to update the code as recommended. Yet, let me explain the reason for such a structure. 

Raising an exception inside an exception will print out following report:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    value = list(value)
TypeError: 'int' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 5, in <module>
    raise AssertionError('blah blah')
AssertionError: blah blah
```

Unlike this, raising exception outside yields just a simple exception record:
```
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    assert value is not None, AssertionError('blah blah')
AssertionError: blah blah
```

I think the latter case (a single exception) is easier to read quickly especially if it happen to be part of a larger block of text (a build log). At the end, failing one python script does not imply all build scripts in all threads will stop immediately. Yet, I understand that this might not be aligned with the intended use. Thus, I will update the code as recommended.

Araneidae replied:
I see what you mean about the nested exceptions, that's not so great.  On the whole I'd just go for `value = list(value)` and expose that slightly unfriendly error directly to the user.

hanak replied:
Ok. I will do.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+    def _format_constant(self, value):
+        if isinstance(value, Parameter):
+            return f'"{str(value)}"'
+        if isinstance(value, str):
```
Use chained `elif` here and end on
```python
else:
     return str(value)
```

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+    def Validate(self, record, fieldname):
+        context = f'ConstArray@{record}.{fieldname}'
+
+        value = self._sanitize(self._value, context)
```
You should have already called `_sanitize` when `_value` was assigned.  If you rename it to `__value` this is more clearly enforced, and we can check that `__value` is only assigned as the result of calling `_sanitize`.  In other words, we shouldn't need this line.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+        value = self._sanitize(self._value, context)
+
+        # EPICS 7.0.3.1 does not consider "[]" as constant.
+        assert len(value) > 0, \
```
Do this test inside `_sanitize`.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+            return '1' if value else '0'
+        return str(value)
+
+    def Validate(self, record, fieldname):
```
After moving the calls to `_sanitize` up to the point of assignment to `_value` we will end up with an empty `Validate`.  At this point we can at least spend a comment to explain what we're _not_ doing (ie, calling `dbVerify`) and why not, and what we wish we could do...

Araneidae reviewed (changes requested):
Changes are:

1. Rename `_value` to `__value` and treat `ConstArray` as a fixed wrapper, so drop the `value` setter and getters.
2. Some style changes
3. `Validate` becomes empty!

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+    def value(self):
+        """A list of values.
+        """
+        return self._value
```
I am going to update to update the code as recommended. Yet, let me explain the reason.

PEP8 (https://www.python.org/dev/peps/pep-0008/#id47) states following:
* "Use one leading underscore only for non-public methods and instance variables."
* "Generally, double leading underscores should be used only to avoid name conflicts with attributes in classes designed to be subclassed."

However, I understand that consistency across the code is important. Which coding standard is used with epicsdbbuilder?

Araneidae replied:
Well, there's something special about `__` variables: they're pseudo private, which means that users of the class can't (readily) access them.  If you use `__value` and a user tries to write, eg `ConstArray([x]).__value` they'll get an error message.  If we're going to validate on assignment to the `ConstArray` then we don't want users going being our backs and doing their own assignment to `_value`.

hanak commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,92 @@
+    def _sanitize(self, raw_value, context_name):
+        try:
+            value_list = list(raw_value)
+        except TypeError:
```
I understand and I agree. Yet, let me add my motivation.

If we leave it up to `list()`, the user will get `'int' object is not iterable` if one attempts to assign an integer instead of anything else. Unless the user knows internals of epicsdbbuilder, the user needs to read corresponding source files. Alternatively, the user will keep a list of error messages and their interprettation. At the same time, epicsdbbuilder knows potential cause of the situation. It can help or guide the user a little bit so that the user can focus on the task rather then on the tool in majority of cases.

Would you like me to remove this try-except?

hanak replied:
I think you have responded to my question in one of your later responses. I will update the code accordingly.

hanak commented:
I missed the `if-elif-else` comment. Just a second, please.

hanak commented:
I have updated the code as recommended. Would it be possible to take a look, please?

Araneidae commented:
Will do.  Looks good, but give me a few days to mull.

hanak commented:
Sure. Thank you.

Araneidae commented:
You could improve the documentation, which is a bit perfunctory at the moment.  I'd cover the following points (concisely):

- This function is used to wrap list of numbers or strings
- The value can only be usefully assigned to certain field types (I forget the types, but you did say upstream), so list these in the documentation
- Note that at present this type is _not_ checked for validity, and so assigning to the wrong field type can result in an unloadable database.

hanak commented:
Sure. Are you referring to `api.rst` or codedoc or both?

Araneidae commented:
> Sure. Are you referring to `api.rst` or codedoc or both?

User level documentation, `api.rst`, please.  I'd make sure you at least copy the `ConstArray` docstring and cover a few more relevant pints.

hanak commented:
I have extended document with description and warning about EPICS inconsistency.

Araneidae commented on docs/api.rst:
```diff
@@ -217,6 +217,47 @@ Helper Functions
+      stringout, lso, lsi, printf, waveform, subArray, and aai.
+      Any other use is undefined and a warning may or may not appear
+      while loading the DB
+      (e.g. assigning ["1.23"] to INP of the record type ai will print a warning
```
Put `` ` `` markup around constant strings `["1.23"]` and `[1.23]`.

Also, 80 column line wrapping, please!  (Yes, I _did_ program on punched cards at university.)

hanak replied:
I am aware I keep running into it. I am sorry. I understand the reason for 80 columns especially when reviewing code with side-by-side view. I have set a vertical column mark to help me to see it clearly.

Araneidae replied:
NP 😛

Araneidae commented on docs/api.rst:
```diff
@@ -217,6 +217,47 @@ Helper Functions
+    * No field type or record type check.
+      ConstArray can be assigned to any field despite the field
+      or the record type does not support Constant Link Values.
+      Use it with link fiels (e.g. INP) of record types stringin,
```
"fiels" should be "fields"

Probably best to highlight the record type names (it'll be enough to use `` `...` `` markup for this), ie write:
```
Use it with link fields (e.g. `INP`) of record types `stringin`,
`stringout`, ...
```

hanak replied:
Typos will be corrected. Markup will be used as suggested.

Araneidae commented on epicsdbbuilder/const_array.py:
```diff
@@ -0,0 +1,87 @@
+                f'ConstArray: expects a string or parameter as element' \
+                f' but an element at the index {index} is {type(value)}.'
+
+            if isinstance(value, (Parameter, str)):
```
There's a small bug here!  A `Parameter` will be of unpredictable type, so you can't treat it as either a string _or_ a number, so I think you need a third test, perhaps along the lines of:
```python
if isinstance(value, Parameter):
    pass
elif isinstance(value, str):
    ... test for strings...
else:
    ... test for numbers...
```
Or else: rename `valid_types` to `number_types` and take out `Parameter`, `str` from the list, and test for that in the last clause:
```python
if isinstance(value, Parameter):
    pass  # Always accept Parameters
elif isinstance(value, str):
    ... string ...
elif isintance(value, number_types):
    ... number ...
else:
    assert False, ... expects a string number or parameter ...
```

hanak replied:
I am afraid I can treat it as unpredictable until certain point. And that point is formatting. Right now, the parameter will be treated as a string (e.g. `"$(X)"`) so that it can be used with string record types (e.g. lsi). If we take it out of the condition, we might end up with `[1, "$(X)", 2]` on the output. This will not work right in case of non-string types (see table below). A workaround could be to treat a parameter as a number and allow the user to supply it in a form of a string (e.g. `$(X)`) if it is meant to be treated as a string. Yet, I am not sure how we can tell a reference to a parameter X apart from a string `$(X)` without additional information. This assume we want to treat a string as a string without any ambiguity.

I think, there is a chance to archive reasonable behavior with some guesswork. Yet, this is case-by-case situation. In case of string record type (lsi/lso, stringin/stringout), a parameter is meant to be a string. Incase of arrays, another field might give a hint. While this could be created, it will not be that quick. Therefore, I focused on getting some functionality without any loose ends or gaps first. Therefore, I added more restrictive rule (i.e. parameters are strings). Definitely, we can remove the support for parameters if we want to. What behavior would you like to prefer?

Here are some results of experimenting with various record types on EPICS 7.0.3.1. Given the style EPICS treats input, it is difficult to tell the intention from an undefined or unintended behavior (a side-effect). The table below shows a record type with relevant fields, a content of the field INP, and result of caget:
| record type | INP | caget | loading DB |
| --- | --- | --- | --- |
| waveform (FTVL=DOUBLE) | `[1, 2, 3]` | `3 1 2 3` |  |
| waveform (FTVL=DOUBLE) | `["1", "2", "3"]` | `3 0 0 0` | warning `dbConvertJSON: String provided, numeric value(s) expected` |
| waveform (FTVL=DOUBLE) | `[1, "2", 3]` | `3 0 0 0` | warning `dbConvertJSON: String provided, numeric value(s) expected` |
| waveform (FTVL=STRING) | `["1", "2", "3"]` | `3 1 2 3` |  |
| waveform (FTVL=STRING) | `[1, "2", 3]` | `3 1 2 3` |  |
| lsi | `[123]` | `` (empty string) |  |
| lsi | `["$(X)"]` | `ValueOfX` |  |
| lsi | `[$(X)]` | `` (empty string) | syntax error while loading |

Araneidae replied:
What does the db loader do when you give it `[1, $(X), 3]` with `X=2` specified?  (I don't have an up to date version of EPICS to hand to test with).  I would have expected that to expand to `[1, 2, 3]`.

However, I see the problem when it comes to expanding the parameter itself: at the moment you treat it as a string and expand it to `"$(X)"`; without knowing the underlying record type (which we can't rely on knowing, particularly if it depends on the value of another field) it's impossible to reliably choose between `$(X)` and `"$(X)"`!

Hmph.  Guess we can't fix this easily, and I'm not sure that it would be a problem in most applications.

hanak replied:
In case of `X=2` and `[1, $(X), 2]`, you will get `[1, 2, 3]`.

Regarding `$(X)` and `"$(X)"`. I think the key would be to identify when the decision is not possible to be made. In such a scenario, the routine should fail and the user should be allowed to add a hint (e.g. `ConstArray([AsString(par)])`). As long as the logging is right, the user will not be frustrated: the log will help the user to move forward without googling or guessing. Yet, I see this as a topic for another pull request. Right now, we can either decide on how we want to treat the parameter (and add all the checks to make sure the we generate valid output) or deny use of parameters. I think I use a parameter with a record type lsi. Thus, I preferred treating the parameter as a string.

Araneidae replied:
Yes, I agree with all that, except I think I'd add the type hint as an optional extra parameter to `ConstArray`, eg `ConstArray([str_par], as_string=True)` and `ConstArray([num_par], as_string=False)` ... and then we can worry whether the parameter should be `as_string`, `as_num`, or `type`, and whether the keyword is optional.

As you say, let's do this another time.

Araneidae reviewed (changes requested):
Nearly there!

Araneidae commented on docs/api.rst:
```diff
@@ -217,6 +217,47 @@ Helper Functions
+      Always refer to EPICS Release Notes (section "Constant Link Values").
+
+
+    Example: PY Source::
```
And the example below, of course

hanak replied:
I will update the document accordingly.

Araneidae commented:
Thank you for your contribution.
